frontend/catalyst/api_extensions/control_flow.py

```diff
@@ -983,35 +1081,69 @@ class WhileLoop(HybridOp):
 
     def trace_quantum(self, ctx, device, trace, qrp) -> QRegPromise:
         cond_trace = self.regions[0].trace
-        res_classical_tracers = self.regions[0].res_classical_tracers
+        expansion_strategy = self.expansion_strategy
         with EvaluationContext.frame_tracing_context(ctx, cond_trace):
+            arg_expanded_classical_tracers = expand_args(
+                self.regions[0].arg_classical_tracers, expansion_strategy=expansion_strategy
+            )[0]
+            res_expanded_classical_tracers, out_type = expand_results(
+                [],
+                arg_expanded_classical_tracers,
+                self.regions[0].res_classical_tracers,
+                expansion_strategy=expansion_strategy,
+            )
```

Comment:
I know I am asking to document a lot of things, but this area of code is one I am not too familiar. And I want to be able to better understand some things. So for example, why are the constants empty in this case and when would they not be empty?

```suggestion
            res_classical_tracers = region.res_classical_tracers
            constants = []
            res_expanded_classical_tracers, out_type = expand_results(
                constants,
                arg_expanded_classical_tracers,
                res_classical_tracers,
                expansion_strategy=expansion_strategy,
            )
```

Reply:
`git grep` reveals that constants is always an empty list.

```
frontend/catalyst/api_extensions/control_flow.py:                res_expanded_tracers, out_type = expand_results(
frontend/catalyst/api_extensions/control_flow.py-                    [],
frontend/catalyst/api_extensions/control_flow.py-                    arg_expanded_classical_tracers,
--
frontend/catalyst/api_extensions/control_flow.py:        out_expanded_classical_tracers = expand_results(
frontend/catalyst/api_extensions/control_flow.py-            [],
frontend/catalyst/api_extensions/control_flow.py-            in_expanded_classical_tracers,
--
frontend/catalyst/api_extensions/control_flow.py:            res_expanded_tracers, _ = expand_results(
frontend/catalyst/api_extensions/control_flow.py-                [],
frontend/catalyst/api_extensions/control_flow.py-                arg_expanded_tracers,
--
frontend/catalyst/api_extensions/control_flow.py:        out_expanded_classical_tracers = expand_results(
frontend/catalyst/api_extensions/control_flow.py-            [],
frontend/catalyst/api_extensions/control_flow.py-            in_expanded_tracers,
--
frontend/catalyst/api_extensions/control_flow.py:            res_expanded_classical_tracers, out_type = expand_results(
frontend/catalyst/api_extensions/control_flow.py-                [],
frontend/catalyst/api_extensions/control_flow.py-                arg_expanded_classical_tracers,
--
frontend/catalyst/api_extensions/control_flow.py:            res_expanded_tracers, out_type = expand_results(
frontend/catalyst/api_extensions/control_flow.py-                [],
frontend/catalyst/api_extensions/control_flow.py-                arg_expanded_tracers,
--
frontend/catalyst/api_extensions/control_flow.py:        out_expanded_classical_tracers = expand_results(
frontend/catalyst/api_extensions/control_flow.py-            [],
frontend/catalyst/api_extensions/control_flow.py-            in_expanded_tracers,

```

Reply:
Same for `expand_args` but we never care about the type being returned:

```
frontend/catalyst/api_extensions/control_flow.py:            *expand_args(
frontend/catalyst/api_extensions/control_flow.py-                aux_tracers + collapse(in_sig.in_type, in_sig.in_expanded_args),
frontend/catalyst/api_extensions/control_flow.py-                expansion_strategy=self.expansion_strategy,
frontend/catalyst/api_extensions/control_flow.py-            )[0],
--
frontend/catalyst/api_extensions/control_flow.py:            arg_expanded_tracers = expand_args(
frontend/catalyst/api_extensions/control_flow.py-                self.regions[0].arg_classical_tracers + [qreg_in],
frontend/catalyst/api_extensions/control_flow.py-                expansion_strategy=expansion_strategy,
frontend/catalyst/api_extensions/control_flow.py-            )[0]
--
frontend/catalyst/api_extensions/control_flow.py:            *expand_args(op.in_classical_tracers, expansion_strategy=expansion_strategy)[0],
frontend/catalyst/api_extensions/control_flow.py-            qrp.actualize(),
frontend/catalyst/api_extensions/control_flow.py-        ]
frontend/catalyst/api_extensions/control_flow.py-
--
frontend/catalyst/api_extensions/control_flow.py:            arg_expanded_classical_tracers = expand_args(
frontend/catalyst/api_extensions/control_flow.py-                self.regions[0].arg_classical_tracers, expansion_strategy=expansion_strategy
frontend/catalyst/api_extensions/control_flow.py-            )[0]
frontend/catalyst/api_extensions/control_flow.py-            res_expanded_classical_tracers, out_type = expand_results(
--
frontend/catalyst/api_extensions/control_flow.py:            arg_expanded_tracers = expand_args(
frontend/catalyst/api_extensions/control_flow.py-                self.regions[1].arg_classical_tracers + [qreg_in],
frontend/catalyst/api_extensions/control_flow.py-                expansion_strategy=expansion_strategy,
frontend/catalyst/api_extensions/control_flow.py-            )[0]
--
frontend/catalyst/api_extensions/control_flow.py:            *expand_args(self.in_classical_tracers, expansion_strategy=expansion_strategy)[0],
frontend/catalyst/api_extensions/control_flow.py-            qrp.actualize(),
frontend/catalyst/api_extensions/control_flow.py-        ]
frontend/catalyst/api_extensions/control_flow.py-
```

Reply:
Can we just remove them?

Reply:
@grwlf this was marked resolved, but I don't think it has been addressed.

Reply:
@erick-xanadu sorry, it was by mistake I think. Removed `constants` from `expand_results`. But for `expand_args` - [check the counterexample](https://github.com/PennyLaneAI/catalyst/blob/55a0ca5a7164b346175661265b211d670888f3f1/frontend/catalyst/jax_extras/tracing.py#L429).

Reply:
Ah it probably closed itself when I merged the suggestion